pkg/cli/oauth.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"math/rand"
+	"net/http"
+	"net/url"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+)
+
+func createServer(ch chan string, state string, port string) *http.Server {
+	return &http.Server{
+		Addr:    ":" + port,
+		Handler: http.HandlerFunc(handler(ch, state)),
+	}
+}
+
+func handler(ch chan string, randState string) func(http.ResponseWriter, *http.Request) {
+	return func(rw http.ResponseWriter, req *http.Request) {
+		if req.URL.Path == "/favicon.ico" {
+			http.Error(rw, "error: visiting /favicon.ico", 404)
+			return
+		}
+		if req.FormValue("state") != randState {
+			log.Printf("state: %s doesn't match. (expected: %s)", req.FormValue("state"), randState)
+			http.Error(rw, "invalid state", 500)
+			return
+		}
+		if code := req.FormValue("code"); code != "" {
+			fmt.Fprintf(rw, "<h1>Success</h1>Authorized.")
+			rw.(http.Flusher).Flush()
+			ch <- code
+			return
+		}
+		ch <- ""
+		http.Error(rw, "invalid code", 500)
+	}
+}
+
+func oauthAccessToken(credential []byte) (*oauth2.Token, error) {
+	oauthConfig, err := google.ConfigFromJSON(credential, "https://www.googleapis.com/auth/androidbuild.internal")
+	if err != nil {
+		return nil, err
+	}
+	redirectURL, err := url.Parse(oauthConfig.RedirectURL)
+	if err != nil {
+		return nil, fmt.Errorf("parse redirect url error: %w", err)
+	}
+	if redirectURL.Hostname() != "localhost" {
+		return nil, errors.New("the redirect url should be `http://localhost:<whatever_port>`")
+	}
+	port := redirectURL.Port()
+	if port == "" {
+		port = "80"
+	}
```

Comment:
Why are we setting a default port rather than returning an error here?

Reply:
Setting the OAuth redirect URI to http://localhost in GCP by default implies port 80. I think there might not be an error.

However, I was hesitant for a while because listening to port 80 requires sudo privileges, as far as I know.

Do you think it's better to simply return an error?

Reply:
Yes, you are correct. Let's return an error here.

Reply:
Okay, fixed.